File: bindings/dotnet/Xui/ControlStyling.cs
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using System.Text;

namespace Xui;

public enum StyleFontStyle : uint { Normal, Italic, Oblique }
public enum StyleAlignment : uint { Start, Center, End, Stretch }

public sealed record PartStyleValues
{
    public ThemeColor? Background { get; init; }
    public ThemeColor? Foreground { get; init; }
    public ThemeColor? BorderBrush { get; init; }
    public Insets? BorderThickness { get; init; }
    public Insets? Padding { get; init; }
    public float? CornerRadius { get; init; }
    public float? Size { get; init; }
    public string? FontFamily { get; init; }
    public float? FontSize { get; init; }
    public uint? FontWeight { get; init; }
    public StyleFontStyle? FontStyle { get; init; }
    public StyleAlignment? HorizontalAlignment { get; init; }
    public StyleAlignment? VerticalAlignment { get; init; }
    public float? Spacing { get; init; }
    public float? RowHeight { get; init; }
    public float? HeaderHeight { get; init; }
    public float? Indentation { get; init; }
    public float? Thickness { get; init; }
    public float? Width { get; init; }
    public float? Height { get; init; }
    public float? RowGap { get; init; }
    public float? ColumnGap { get; init; }
    public uint? MaximumLines { get; init; }
    public bool? Wrapping { get; init; }
    internal static readonly PartStyleValues Empty = new();

    internal void Validate(StyleTarget target, StylePart part, ulong state = 0)
    {
        if (!Enum.IsDefined(target) || !Enum.IsDefined(part))
            throw new ArgumentException("Unsupported style target or part.");
        ulong mask = (Background.HasValue ? 1u : 0) | (Foreground.HasValue ? 2u : 0) |
            (BorderBrush.HasValue ? 4u : 0) | (BorderThickness.HasValue ? 8u : 0) |
            (Padding.HasValue ? 16u : 0) | (CornerRadius.HasValue ? 32u : 0) | (Size.HasValue ? 64u : 0) |
            (FontFamily is not null ? 128u : 0) | (FontSize.HasValue ? 256u : 0) | (FontWeight.HasValue ? 512u : 0) |
            (FontStyle.HasValue ? 1024u : 0) | (HorizontalAlignment.HasValue ? 2048u : 0) | (VerticalAlignment.HasValue ? 4096u : 0) |
            (Spacing.HasValue ? 8192u : 0) | (RowHeight.HasValue ? 16384u : 0) | (HeaderHeight.HasValue ? 32768u : 0) |
            (Indentation.HasValue ? 65536u : 0) | (Thickness.HasValue ? 131072u : 0) | (Width.HasValue ? 262144u : 0) |
            (Height.HasValue ? 524288u : 0) | (RowGap.HasValue ? 1048576u : 0) | (ColumnGap.HasValue ? 2097152u : 0) |
            (MaximumLines.HasValue ? 4194304u : 0) | (Wrapping.HasValue ? 8388608u : 0);
        var schema = StyleSchemaCatalog.Find(target, part) ?? throw new ArgumentException("Unsupported style target or part.");
        var limits = StyleSchemaCatalog.Limits(target, part);
        if ((state & ~schema.States) != 0) throw new ArgumentException("Unsupported state on this style part.");
        ulong allowed = state == 0 ? schema.Properties : schema.StateProperties;
        if ((mask & ~allowed) != 0) throw new ArgumentException("Unsupported property on this style part.");
        new ButtonStyleValues { Background = Background, Foreground = Foreground, BorderBrush = BorderBrush,
            BorderThickness = BorderThickness, Padding = Padding, CornerRadius = CornerRadius }.Validate();
        if (Size is { } size) ButtonStyleValues.Dimension(size);
        if (FontFamily is { } family && (family.Length == 0 || family.Length > limits.FontFamilyUtf16 || family.Contains('\0') ||
            new UTF8Encoding(false, true).GetByteCount(family) > 1024)) throw new ArgumentException("Invalid font family.");
        if (FontSize is { } fontSize) { ButtonStyleValues.Dimension(fontSize); if (fontSize == 0 || fontSize > limits.FontSize) throw new ArgumentOutOfRangeException(nameof(FontSize)); }
        if (FontWeight is < 1 or > 999) throw new ArgumentOutOfRangeException(nameof(FontWeight));
        if (FontStyle is { } fontStyle && (!Enum.IsDefined(fontStyle) || (limits.FontStyles & (1u << (int)fontStyle)) == 0))
            throw new ArgumentOutOfRangeException(nameof(FontStyle));
        if (HorizontalAlignment is { } horizontal && (!Enum.IsDefined(horizontal) || (limits.HorizontalAlignments & (1u << (int)horizontal)) == 0))
            throw new ArgumentOutOfRangeException(nameof(HorizontalAlignment));
        if (VerticalAlignment is { } vertical && (!Enum.IsDefined(vertical) || (limits.VerticalAlignments & (1u << (int)vertical)) == 0))
            throw new ArgumentOutOfRangeException(nameof(VerticalAlignment));
        foreach (var dimension in new[] { Spacing, RowHeight, HeaderHeight, Indentation, Thickness, Width, Height, RowGap, ColumnGap })
            if (dimension is { } number) ButtonStyleValues.Dimension(number);
        if (RowHeight is 0) throw new ArgumentOutOfRangeException(nameof(RowHeight));
        if (MaximumLines is > 32768) throw new ArgumentOutOfRangeException(nameof(MaximumLines));
    }
    internal unsafe void Append(List<Native.StyleProperty> records, StylePart part, ulong state, Window.Pins pins)
    {
        if (FontFamily is { } fontFamily) _ = new UTF8Encoding(false, true).GetByteCount(fontFamily);
        Native.StyleProperty Record(uint property, uint type) => new() {
            Size = (uint)sizeof(Native.StyleProperty), Version = 0x10000,
            Property = property, ValueType = type, Part = (uint)part, State = state
        };
        void Color(uint property, ThemeColor? color)
        {
            if (color is not { } c) return;
            var r = Record(property, 1); r.Color = new() { Light = c.Light, Dark = c.Dark }; records.Add(r);
        }
        void Edges(uint property, Insets? insets)
        {
            if (insets is not { } e) return;
            var r = Record(property, 2); r.Insets = new() { Left = e.Left, Top = e.Top, Right = e.Right, Bottom = e.Bottom }; records.Add(r);
        }
        void Number(uint property, double? number)
        {
            if (number is not { } n) return;
            var r = Record(property, 3); r.Number = n; records.Add(r);
        }
        Color(1, Background); Color(2, Foreground); Color(4, BorderBrush);
        Edges(8, BorderThickness); Edges(16, Padding); Number(32, CornerRadius); Number(64, Size);
        if (FontFamily is { } family) { var r = Record(128, 4); r.Text = pins.Text(family); records.Add(r); }
        Number(256, FontSize); Number(512, FontWeight); Number(1024, FontStyle is { } fs ? (uint)fs : null);
        Number(2048, HorizontalAlignment is { } h ? (uint)h : null); Number(4096, VerticalAlignment is { } v ? (uint)v : null);
        Number(8192, Spacing); Number(16384, RowHeight); Number(32768, HeaderHeight); Number(65536, Indentation);
        Number(131072, Thickness); Number(262144, Width); Number(524288, Height);
        Number(1048576, RowGap); Number(2097152, ColumnGap); Number(4194304, MaximumLines);
        Number(8388608, Wrapping is { } wrap ? wrap ? 1 : 0 : null);
    }
    internal static unsafe PartStyleValues FromNative(IEnumerable<Native.StyleProperty> records)
    {
        var result = Empty;
        foreach (var r in records)
        {
            var c = new ThemeColor(r.Color.Light, r.Color.Dark);
            var e = new Insets(r.Insets.Left, r.Insets.Top, r.Insets.Right, r.Insets.Bottom);
            result = r.Property switch {
                1 => result with { Background = c }, 2 => result with { Foreground = c },
                4 => result with { BorderBrush = c }, 8 => result with { BorderThickness = e },
                16 => result with { Padding = e }, 32 => result with { CornerRadius = (float)r.Number },
                64 => result with { Size = (float)r.Number },
                128 => result with { FontFamily = Marshal.PtrToStringUTF8((nint)r.Text.Data, checked((int)r.Text.Length)) },
                256 => result with { FontSize = (float)r.Number }, 512 => result with { FontWeight = (uint)r.Number },
                1024 => result with { FontStyle = (StyleFontStyle)(uint)r.Number },
                2048 => result with { HorizontalAlignment = (StyleAlignment)(uint)r.Number },
                4096 => result with { VerticalAlignment = (StyleAlignment)(uint)r.Number },
                8192 => result with { Spacing = (float)r.Number }, 16384 => result with { RowHeight = (float)r.Number },
                32768 => result with { HeaderHeight = (float)r.Number }, 65536 => result with { Indentation = (float)r.Number },
                131072 => result with { Thickness = (float)r.Number }, 262144 => result with { Width = (float)r.Number },
                524288 => result with { Height = (float)r.Number }, 1048576 => result with { RowGap = (float)r.Number },
                2097152 => result with { ColumnGap = (float)r.Number }, 4194304 => result with { MaximumLines = (uint)r.Number },
                8388608 => result with { Wrapping = r.Number != 0 },
                _ => throw new InvalidOperationException("Unknown native style property.")
            };
        }
        return result;
    }
}

public sealed record PartStyle(StylePart Part, PartStyleValues Values);
public sealed record ControlStyleRule(StylePart Part, StyleState State, PartStyleValues Values);

/// <summary>An immutable shared style with weak native identities for each window.</summary>
public sealed class ControlStyle
{
    private sealed class NativeIdentity { internal ulong Value; }
    private readonly ConditionalWeakTable<Window, NativeIdentity> identities = new();
    private readonly int depth;
    public StyleTarget Target { get; }
    public IReadOnlyList<PartStyle> Parts { get; }
    public IReadOnlyList<ControlStyleRule> Rules { get; }
    public ControlStyle? BasedOn { get; }
    public ControlStyle(StyleTarget target, IReadOnlyList<PartStyle> parts,
        IReadOnlyList<ControlStyleRule>? rules = null, ControlStyle? basedOn = null)
    {
        ArgumentNullException.ThrowIfNull(parts);
        if (!Enum.IsDefined(target) || !StyleSchemaCatalog.HasTarget(target)) throw new ArgumentOutOfRangeException(nameof(target));
        if (basedOn is not null && basedOn.Target != target) throw new ArgumentException("Style base target mismatch.");
        depth = (basedOn?.depth ?? 0) + 1;
        if (depth > 16 || parts.Count > 64 || rules?.Count > 256) throw new ArgumentException("Style limits exceeded.");
        var partCopy = parts.ToArray(); var ruleCopy = rules?.ToArray() ?? [];
        var keys = new HashSet<(StylePart, ulong)>();
        foreach (var p in partCopy)
        {
            ArgumentNullException.ThrowIfNull(p); ArgumentNullException.ThrowIfNull(p.Values);
            p.Values.Validate(target, p.Part);
            if (!keys.Add((p.Part, 0))) throw new ArgumentException("Duplicate style part.");
        }
        foreach (var r in ruleCopy)
        {
            ArgumentNullException.ThrowIfNull(r); ArgumentNullException.ThrowIfNull(r.Values);
            if (!Enum.IsDefined(r.State)) throw new ArgumentException("A rule requires one supported state.");
            r.Values.Validate(target, r.Part, (ulong)r.State);
            if (!keys.Add((r.Part, (ulong)r.State))) throw new ArgumentException("Duplicate style state for this part.");
        }
        Target = target; Parts = Array.AsReadOnly(partCopy); Rules = Array.AsReadOnly(ruleCopy); BasedOn = basedOn;
    }
    internal unsafe void WithNativeHandle(Window window, Action<ulong> apply)
    {
        window.Guard();
        var identity = identities.GetOrCreateValue(window);
        if (identity.Value != 0)
        {
            ulong retained = 0;
            window.Check(Native.ControlStyleReacquire(window.Handle, identity.Value, &retained));
            if (retained != 0)
            {
                try { apply(retained); } finally { window.Check(Native.ControlStyleRelease(retained)); }
                return;
            }
        }
        var records = new List<Native.StyleProperty>();
        using var pins = new Window.Pins();
        foreach (var p in Parts) p.Values.Append(records, p.Part, 0, pins);
        foreach (var r in Rules) r.Values.Append(records, r.Part, (ulong)r.State, pins);
        var array = records.ToArray();
        void Create(ulong parent)
        {
            ulong handle = 0;
            fixed (Native.StyleProperty* p = array)
            {
                var options = new Native.ControlStyleOptions { Size = (uint)sizeof(Native.ControlStyleOptions),
                    Version = 0x10000, Target = (uint)Target, Properties = p, PropertyCount = (uint)array.Length, BasedOn = parent };
                window.Check(Native.ControlStyleCreate(window.Handle, &options, &handle));
            }
            identity.Value = handle;
            try { apply(handle); } finally { window.Check(Native.ControlStyleRelease(handle)); }
        }
        if (BasedOn is { } parent) parent.WithNativeHandle(window, Create);
        else Create(0);
    }
    internal unsafe bool TryApply(Window window, ulong handle, bool tooltip = false)
    {
        if (!identities.TryGetValue(window, out var identity)) return false;
        uint applied = 0;
        window.Check(tooltip ? Native.WindowTrySetTooltipStyle(handle, identity.Value, &applied) :
            Native.ControlTrySetStyle(handle, identity.Value, &applied));
        return applied != 0;
    }
}

public sealed unsafe partial class Window
{
    private ControlStyle? tooltipStyle;
    public ControlStyle? TooltipStyle { get { Guard(); return tooltipStyle; } set => SetTooltipStyle(value); }
    public Window SetTooltipStyle(ControlStyle? value)
    {
        Guard();
        if (value is null) Check(Native.WindowSetTooltipStyle(Handle, 0));
        else
        {
            if (value.Target != StyleTarget.Tooltip) throw new ArgumentException("Expected a Tooltip style.", nameof(value));
            if (!value.TryApply(this, Handle, tooltip: true))
                value.WithNativeHandle(this, handle => Check(Native.WindowSetTooltipStyle(Handle, handle)));
        }
        tooltipStyle = value;
        return this;
    }
    public Window SetTooltipStyleValues(StylePart part, PartStyleValues values)
    {
        Guard(); ArgumentNullException.ThrowIfNull(values);
        values.Validate(StyleTarget.Tooltip, part);
        using var pins = new Pins();
        var records = new List<Native.StyleProperty>();
        values.Append(records, part, 0, pins);
        var array = records.ToArray();
        fixed (Native.StyleProperty* p = array)
            Check(Native.WindowSetTooltipStyleValues(Handle, (uint)part, p, (uint)array.Length));
        return this;
    }
    public PartStyleValues GetTooltipStyleValues(StylePart part, bool effective = false)
    {
        Guard();
        uint count = 0;
        Check(Native.WindowGetTooltipStyleValues(Handle, (uint)part, effective ? 1u : 0u, null, 0, &count));
        var records = new Native.StyleProperty[count];
        fixed (Native.StyleProperty* p = records)
            Check(Native.WindowGetTooltipStyleValues(Handle, (uint)part, effective ? 1u : 0u, p, count, &count));
        return PartStyleValues.FromNative(records);
    }
}

public abstract unsafe partial class Element
{
    private ControlStyle? controlStyle;
    public ControlStyle? ControlStyle { get { Window.Guard(); return controlStyle; } set => SetControlStyle(value); }
    public Element SetControlStyle(ControlStyle? value)
    {
        Window.Guard();
        if (value is null) Window.Check(Native.ControlSetStyle(Handle, 0));
        else if (!value.TryApply(Window, Handle))
            value.WithNativeHandle(Window, handle => Window.Check(Native.ControlSetStyle(Handle, handle)));
        controlStyle = value;
        return this;
    }
    public Element SetControlStyleValues(StylePart part, PartStyleValues values)
    {
        Window.Guard(); ArgumentNullException.ThrowIfNull(values);
        using var pins = new Window.Pins();
        var records = new List<Native.StyleProperty>(); values.Append(records, part, 0, pins);
        var array = records.ToArray();
        fixed (Native.StyleProperty* p = array)
            Window.Check(Native.ControlSetStyleValues(Handle, (uint)part, p, (uint)array.Length));
        return this;
    }
    public PartStyleValues GetControlStyleValues(StylePart part, bool effective = false)
    {
        Window.Guard();
        uint count = 0;
        Window.Check(Native.ControlGetStyleValues(Handle, (uint)part, effective ? 1u : 0u, null, 0, &count));
        var records = new Native.StyleProperty[count];
        fixed (Native.StyleProperty* p = records)
            Window.Check(Native.ControlGetStyleValues(Handle, (uint)part, effective ? 1u : 0u, p, count, &count));
        return PartStyleValues.FromNative(records);
    }
}

public sealed partial class Toggle
{
    public ControlStyle? Style { get => ControlStyle; set => SetControlStyle(value); }
    public Toggle SetStyle(ControlStyle? value) { SetControlStyle(value); return this; }
    public Toggle SetStyleValues(StylePart part, PartStyleValues values) { SetControlStyleValues(part, values); return this; }
    public PartStyleValues GetStyleValues(StylePart part, bool effective = false) => GetControlStyleValues(part, effective);
}

File: bindings/rust/xui/src/control_styling.rs
use super::*;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
#[repr(u32)]
pub enum StyleTarget {
    Toggle,
}
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
#[repr(u32)]
pub enum StylePart {
    Root,
    Label,
    Indicator,
    Mark,
}
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
#[repr(u64)]
pub enum StyleState {
    Focused = 1,
    Checked = 2,
    Hovered = 4,
    Pressed = 8,
    Disabled = 16,
}

#[derive(Clone, Copy, Debug, Default, PartialEq)]
pub struct PartStyleValues {
    pub background: Option<ThemeColor>,
    pub foreground: Option<ThemeColor>,
    pub border_brush: Option<ThemeColor>,
    pub border_thickness: Option<Insets>,
    pub padding: Option<Insets>,
    pub corner_radius: Option<f32>,
    pub size: Option<f32>,
}
impl PartStyleValues {
    fn append(
        self,
        part: StylePart,
        state: u64,
        output: &mut Vec<sys::StyleProperty>,
    ) -> Result<()> {
        let allowed = match part {
            StylePart::Root => 63,
            StylePart::Indicator => 109,
            _ => 2,
        };
        let mut record = |property, value_type, color, insets, number| -> Result<()> {
            if property & allowed == 0 {
                return Err(invalid("Unsupported property on this style part."));
            }
            output.push(sys::StyleProperty {
                size: size_of::<sys::StyleProperty>() as u32,
                version: 0x10000,
                property,
                value_type,
                part: part as u32,
                state,
                color,
                insets,
                number,
                ..Default::default()
            });
            Ok(())
        };
        for (property, value) in [
            (1, self.background),
            (2, self.foreground),
            (4, self.border_brush),
        ] {
            if let Some(c) = value {
                if c.light > 0xffffff || c.dark > 0xffffff {
                    return Err(invalid("Colors must be RGB24."));
                }
                record(
                    property,
                    1,
                    sys::ThemeColor {
                        light: c.light,
                        dark: c.dark,
                    },
                    Default::default(),
                    0.0,
                )?;
            }
        }
        fn dimension(value: f32) -> Result<()> {
            if !value.is_finite() || !(0.0..=32768.0).contains(&value) {
                return Err(invalid(
                    "Style dimensions must be finite DIPs from zero through 32768.",
                ));
            }
            Ok(())
        }
        for (property, value) in [(8, self.border_thickness), (16, self.padding)] {
            if let Some(e) = value {
                for v in [e.left, e.top, e.right, e.bottom] {
                    dimension(v)?;
                }
                record(
                    property,
                    2,
                    Default::default(),
                    sys::StyleInsets {
                        left: e.left,
                        top: e.top,
                        right: e.right,
                        bottom: e.bottom,
                    },
                    0.0,
                )?;
            }
        }
        for (property, value) in [(32, self.corner_radius), (64, self.size)] {
            if let Some(v) = value {
                dimension(v)?;
                record(
                    property,
                    3,
                    Default::default(),
                    Default::default(),
                    v as f64,
                )?;
            }
        }
        Ok(())
    }
    fn from_native(records: &[sys::StyleProperty]) -> Result<Self> {
        let mut result = Self::default();
        for r in records {
            let c = ThemeColor::new(r.color.light, r.color.dark);
            let e = Insets {
                left: r.insets.left,
                top: r.insets.top,
                right: r.insets.right,
                bottom: r.insets.bottom,
            };
            match r.property {
                1 => result.background = Some(c),
                2 => result.foreground = Some(c),
                4 => result.border_brush = Some(c),
                8 => result.border_thickness = Some(e),
                16 => result.padding = Some(e),
                32 => result.corner_radius = Some(r.number as f32),
                64 => result.size = Some(r.number as f32),
                _ => return Err(invalid("Unknown native style property.")),
            }
        }
        Ok(result)
    }
}
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct PartStyle {
    pub part: StylePart,
    pub values: PartStyleValues,
}
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct ControlStyleRule {
    pub part: StylePart,
    pub state: StyleState,
    pub values: PartStyleValues,
}
struct ControlStyleDefinition {
    target: StyleTarget,
    parts: Vec<PartStyle>,
    rules: Vec<ControlStyleRule>,
    based_on: Option<ControlStyle>,
    depth: usize,
    identities: RefCell<Vec<(Weak<Inner>, u64)>>,
}
/// Immutable values, shared between controls without retaining their windows.
#[derive(Clone)]
pub struct ControlStyle(Rc<ControlStyleDefinition>);
impl ControlStyle {
    pub fn new(
        target: StyleTarget,
        parts: &[PartStyle],
        rules: &[ControlStyleRule],
        based_on: Option<&Self>,
    ) -> Result<Self> {
        let depth = based_on.map_or(1, |b| b.0.depth + 1);
        if depth > 16 || parts.len() > 64 || rules.len() > 256 {
            return Err(invalid("Control style limits exceeded."));
        }
        if based_on.is_some_and(|b| b.0.target != target) {
            return Err(invalid("Style base target mismatch."));
        }
        let mut keys = Vec::new();
        let mut records = Vec::new();
        for p in parts {
            if keys.contains(&(p.part, 0)) {
                return Err(invalid("Duplicate style part."));
            }
            keys.push((p.part, 0));
            p.values.append(p.part, 0, &mut records)?;
        }
        for r in rules {
            if keys.contains(&(r.part, r.state as u64)) {
                return Err(invalid("Duplicate style state for this part."));
            }
            keys.push((r.part, r.state as u64));
            r.values.append(r.part, r.state as u64, &mut records)?;
        }
        Ok(Self(Rc::new(ControlStyleDefinition {
            target,
            parts: parts.to_vec(),
            rules: rules.to_vec(),
            based_on: based_on.cloned(),
            depth,
            identities: RefCell::new(Vec::new()),
        })))
    }
    pub fn target(&self) -> StyleTarget {
        self.0.target
    }
    pub fn parts(&self) -> &[PartStyle] {
        &self.0.parts
    }
    pub fn rules(&self) -> &[ControlStyleRule] {
        &self.0.rules
    }
    pub fn based_on(&self) -> Option<&Self> {
        self.0.based_on.as_ref()
    }
    fn identity(&self, owner: &Rc<Inner>) -> Option<u64> {
        let mut identities = self.0.identities.borrow_mut();
        identities.retain(|(window, _)| window.strong_count() != 0);
        identities
            .iter()
            .find(|(window, _)| window.ptr_eq(&Rc::downgrade(owner)))
            .map(|(_, id)| *id)
    }
    fn remember(&self, owner: &Rc<Inner>, identity: u64) {
        let mut identities = self.0.identities.borrow_mut();
        if let Some((_, id)) = identities
            .iter_mut()
            .find(|(window, _)| window.ptr_eq(&Rc::downgrade(owner)))
        {
            *id = identity;
        } else {
            identities.push((Rc::downgrade(owner), identity));
        }
    }
    fn with_native_handle(
        &self,
        owner: &Rc<Inner>,
        apply: &mut dyn FnMut(u64) -> Result<()>,
    ) -> Result<()> {
        if let Some(identity) = self.identity(owner) {
            let mut handle = 0;
            owner.check(unsafe {
                sys::xui_control_style_reacquire(owner.handle, identity, &mut handle)
            })?;
            if handle != 0 {
                let applied = apply(handle);
                let released = owner.check(unsafe { sys::xui_control_style_release(handle) });
                return applied.and(released);
            }
        }
        let mut records = Vec::new();
        for p in &self.0.parts {
            p.values.append(p.part, 0, &mut records)?;
        }
        for r in &self.0.rules {
            r.values.append(r.part, r.state as u64, &mut records)?;
        }
        let mut create = |base| {
            let options = sys::ControlStyleOptions {
                size: size_of::<sys::ControlStyleOptions>() as u32,
                version: 0x10000,
                target: self.0.target as u32,
                properties: records.as_ptr(),
                property_count: records.len() as u32,
                based_on: base,
                ..Default::default()
            };
            let mut handle = 0;
            owner.check(unsafe {
                sys::xui_control_style_create(owner.handle, &options, &mut handle)
            })?;
            self.remember(owner, handle);
            let applied = apply(handle);
            let released = owner.check(unsafe { sys::xui_control_style_release(handle) });
            applied.and(released)
        };
        if let Some(base) = &self.0.based_on {
            base.with_native_handle(owner, &mut create)
        } else {
            create(0)
        }
    }
}
impl Element {
    pub fn set_control_style(&self, style: Option<&ControlStyle>) -> Result<()> {
        let mut apply = |handle| {
            self.owner
                .check(unsafe { sys::xui_control_set_style(self.handle, handle) })
        };
        if let Some(style) = style {
            if let Some(identity) = style.identity(&self.owner) {
                let mut applied = 0;
                self.owner.check(unsafe {
                    sys::xui_control_try_set_style(self.handle, identity, &mut applied)
                })?;
                if applied != 0 {
                    return Ok(());
                }
            }
            style.with_native_handle(&self.owner, &mut apply)
        } else {
            apply(0)
        }
    }
    pub fn set_control_style_values(&self, part: StylePart, values: PartStyleValues) -> Result<()> {
        let mut records = Vec::new();
        values.append(part, 0, &mut records)?;
        self.owner.check(unsafe {
            sys::xui_control_set_style_values(
                self.handle,
                part as u32,
                records.as_ptr(),
                records.len() as u32,
            )
        })
    }
    pub fn control_style_values(
        &self,
        part: StylePart,
        effective: bool,
    ) -> Result<PartStyleValues> {
        let mut count = 0;
        self.owner.check(unsafe {
            sys::xui_control_get_style_values(
                self.handle,
                part as u32,
                effective as u32,
                std::ptr::null_mut(),
                0,
                &mut count,
            )
        })?;
        let mut records = vec![sys::StyleProperty::default(); count as usize];
        self.owner.check(unsafe {
            sys::xui_control_get_style_values(
                self.handle,
                part as u32,
                effective as u32,
                records.as_mut_ptr(),
                count,
                &mut count,
            )
        })?;
        PartStyleValues::from_native(&records)
    }
}
impl Toggle {
    pub fn set_style(&self, style: Option<&ControlStyle>) -> Result<()> {
        self.set_control_style(style)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn generic_style_schema_and_records() -> Result<()> {
        let values = PartStyleValues {
            size: Some(18.0),
            background: Some(ThemeColor::new(0, 0xffffff)),
            ..Default::default()
        };
        let mut records = Vec::new();
        values.append(StylePart::Indicator, 0, &mut records)?;
        assert_eq!(PartStyleValues::from_native(&records)?, values);
        assert_eq!(records[0].color.dark, 0xffffff);
        assert!(values.append(StylePart::Label, 0, &mut Vec::new()).is_err());
        assert!(
            PartStyleValues {
                size: Some(f32::NAN),
                ..Default::default()
            }
            .append(StylePart::Indicator, 0, &mut Vec::new())
            .is_err()
        );
        let part = PartStyle {
            part: StylePart::Indicator,
            values,
        };
        assert!(ControlStyle::new(StyleTarget::Toggle, &[part, part], &[], None).is_err());
        let mut style = ControlStyle::new(StyleTarget::Toggle, &[part], &[], None)?;
        for _ in 1..16 {
            style = ControlStyle::new(StyleTarget::Toggle, &[], &[], Some(&style))?;
        }
        assert!(ControlStyle::new(StyleTarget::Toggle, &[], &[], Some(&style)).is_err());
        Ok(())
    }

    #[test]
    fn generic_toggle_native_lifecycle() -> Result<()> {
        let style = ControlStyle::new(
            StyleTarget::Toggle,
            &[
                PartStyle {
                    part: StylePart::Root,
                    values: PartStyleValues {
                        foreground: Some(0x123456.into()),
                        ..Default::default()
                    },
                },
                PartStyle {
                    part: StylePart::Indicator,
                    values: PartStyleValues {
                        size: Some(18.0),
                        background: Some(0.into()),
                        ..Default::default()
                    },
                },
            ],
            &[ControlStyleRule {
                part: StylePart::Indicator,
                state: StyleState::Checked,
                values: PartStyleValues {
                    background: Some(0x654321.into()),
                    ..Default::default()
                },
            }],
            None,
        )?;
        let window = Window::new("Generic Toggle", 300.0, 200.0)?;
        let toggle = window.toggle("Styled")?;
        toggle.set_style(Some(&style))?;
        assert_eq!(
            toggle.effective_style_values(StylePart::Label)?.foreground,
            Some(0x123456.into())
        );
        toggle.checked(true)?;
        assert_eq!(
            toggle
                .effective_style_values(StylePart::Indicator)?
                .background,
            Some(0x654321.into())
        );
        let local = PartStyleValues {
            foreground: Some(0.into()),
            ..Default::default()
        };
        toggle.set_style_values(StylePart::Root, local)?;
        toggle.set_style(None)?;
        assert_eq!(
            toggle.effective_style_values(StylePart::Label)?.foreground,
            Some(0.into())
        );
        toggle.set_style_values(StylePart::Root, PartStyleValues::default())?;
        assert_eq!(
            toggle.effective_style_values(StylePart::Label)?,
            PartStyleValues::default()
        );
        toggle.set_style(Some(&style))?;
        let marker = window.button("Before")?.handle;
        for _ in 0..256 {
            toggle.set_style(Some(&style))?;
        }
        assert_eq!(window.button("After")?.handle, marker + 1);
        assert!(
            window
                .button("Wrong target")?
                .set_control_style(Some(&style))
                .is_err()
        );
        let other = Window::new("Other", 300.0, 200.0)?;
        let shared = other.toggle("Shared")?;
        shared.set_style(Some(&style))?;
        assert_eq!(
            shared.effective_style_values(StylePart::Indicator)?.size,
            Some(18.0)
        );
        Ok(())
    }
}
impl Toggle {
    pub fn set_style_values(&self, part: StylePart, values: PartStyleValues) -> Result<()> {
        self.set_control_style_values(part, values)
    }
    pub fn style_values(&self, part: StylePart) -> Result<PartStyleValues> {
        self.control_style_values(part, false)
    }
    pub fn effective_style_values(&self, part: StylePart) -> Result<PartStyleValues> {
        self.control_style_values(part, true)
    }
}
